### PR TITLE
update grpc-go version to v1.32.0 which has some breaking api changes

### PR DESCRIPTION
**Pull request by skyao**
update grpc-go version to v1.32.0 which has some breaking api changes in naming and loadbalancer package.

The discussion is in this issue #12124

In this PR, there are two changes according to the breaking api changes in grpc-go (from v1.29.1 to v1.32.0):

1. grpc naming package removed

    For naming resolver, it is hard to change all the code in etcd packages (naming and grpcproxy) from grpc naming to grpc resolver. A simple way is to copy the grpc naming package (in fact just one file naming.go containing 200 lines code) into etcd naming package.

   I put this naming.go file in `clientv3/naming/grpcnaming/naming.go`. 

2. grpc loadbalancer api changed

       For loadbalancer, there are some small breaking api changes but one BIG change in Balancer interface...

      I changed `clientv3/balancer/balancer.go` by following the code of grpc  [`balancer/base/balancer.go`](https://github.com/grpc/grpc-go/blob/506b7730668b5a13465224b0d8133f974a3f843d/balancer/base/balancer.go). 

      Please help to review this file carefully, thanks.

### CODE REVIEW

**Review by ptabor — CHANGES_REQUESTED** [review]
Overall seems go to me. I don't have experience with the grpc balancer code though.
The biggest worry is change of protobuf to 1.4.2 

FYI: The balancer tests were flaky before this PR https://github.com/etcd-io/etcd/issues/12372#issuecomment-705534078

**Comment by ptabor** [comment]
Please make sure:  `PASSES="fmt unit integration" ./test` passes locally. 
It seems there are actionable test/build failures.

File: clientv3/naming/grpc.go [hunk 1]
```diff
@@ -18,11 +18,10 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
-
```
Comment (xiang90) [inline clientv3/naming/grpc.go:21]: @skyao can we mark this package as deprecated? Better, we should move it into the grpcproxy package as an internal implementation detail. 

There is no point to expose this anymore since gRPC already stop supporting it. When our users bump gRPC the same version as the etcd client, they cannot use this package anyway. If they use an older version of gRPC, they should also be using an older version of etcd client with naming support.

So we will lose nothing anyway.

**Comment by ptabor** [comment]
The clientv3 API is documented as API for external users (not pure dependency of grpc proxy): 
> https://etcd.io/docs/v3.4.0/dev-guide/grpc_naming/

Etcd is even mentioned as an (only) example of external resolver: 
  https://github.com/grpc/grpc/blob/master/doc/naming.md

But indeed just preserving it as copy would be useless to external customers of grpc>=1.30 as grpc.WithBalancer(b).
So to me: 
  - we need to rewrite it to the resolver API
  - at the same time we need to use the same JSON for update message (so to copy the struct alone).

How about spiting the PR separately for balancer part vs. naming part ?

File: go.mod [hunk 2]
```diff
@@ -11,8 +11,8 @@ require (
 	github.com/dustin/go-humanize v0.0.0-20171111073723-bb3d318650d4
 	github.com/fatih/color v1.7.0 // indirect
 	github.com/gogo/protobuf v1.3.1
-	github.com/golang/groupcache v0.0.0-20160516000752-02826c3e7903
-	github.com/golang/protobuf v1.3.5
+	github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e
+	github.com/golang/protobuf v1.4.2
```
[the remark this answers is not in the document]
Comment (skyao) [reply go.mod:15]: I checked the grpc code (go.mod file ) that grpc v1.32.0 only requires protobuf 1.3.3: 

```
github.com/golang/protobuf v1.3.3
```

Why it changed to v1.4.2?

**Comment by xiang90** [comment]
@ptabor 

> The clientv3 API is documented as API for external users (not pure dependency of grpc proxy):

It does not really matter. 

The naming package is only for the gRPC-go client to register and resolve endpoints. There are two cases:

1. Project depends on gRPC < 1.32.0. It might depend on the naming interface provided by gRPC. Then they should use etcd client that depends on gRPC < 1.32.0 which has naming support. So it is fine before or after this change.

2. Project depends on gRPC >= 1.32.0. Then they SHOULD NOT depend on naming interface provided by gRPC, since gRPC 1.32.0 already removed the support anyway. They should also use etcd client that depends on gRPC >= 1.32.0 which do not support naming package. So it is also just fine.

File: go.mod (hunk 2 above)
[the remark this answers is not in the document]
Comment (skyao) [reply go.mod]: I have added a commit to downgrade the github.com/golang/protobuf version back to 1.3.5, the lasted release before v1.4.

File: clientv3/balancer/balancer.go [hunk 3]
```diff
@@ -136,15 +136,34 @@ type baseBalancer struct {
...
+
+	resolverErr error // the last error reported by the resolver; cleared on successful resolution
 }
 
-// HandleResolvedAddrs implements "grpc/balancer.Balancer" interface.
-// gRPC sends initial or updated resolved addresses from "Build".
-func (bb *baseBalancer) HandleResolvedAddrs(addrs []resolver.Address, err error) {
-	if err != nil {
-		bb.lg.Warn("HandleResolvedAddrs called with error", zap.String("balancer-id", bb.id), zap.Error(err))
+func (bb *baseBalancer) ResolverError(err error) {
```
[the remark this answers is not in the document]
Comment (skyao) [reply clientv3/balancer/balancer.go:144]: Document added.

**Comment by skyao** [comment]
> The clientv3 API is documented as API for external users (not pure dependency of grpc proxy):
> 
> > https://etcd.io/docs/v3.4.0/dev-guide/grpc_naming/
> 
> Etcd is even mentioned as an (only) example of external resolver:
> https://github.com/grpc/grpc/blob/master/doc/naming.md
> 
> But indeed just preserving it as copy would be useless to external customers of grpc>=1.30 as grpc.WithBalancer(b).
> So to me:
> 
> * we need to rewrite it to the resolver API
> * at the same time we need to use the same JSON for update message (so to copy the struct alone).
> 
> How about spiting the PR separately for balancer part vs. naming part ?

I think we can't split them: if we upgrade the grpc version and only fix one of them, then etcd can't compile.

File: go.mod (hunk 2 above)
[the remark this answers is not in the document]
Comment (skyao) [reply go.mod]: by command go get github.com/golang/protobuf@v1.3.5

[the remark this answers is not in the document]
Comment (skyao) [reply go.mod]: I found that when I execute command "go mod tidy", the version of github.com/golang/protobuf will auto changed to version v1.4.2.

File: go.mod [hunk 4]
```diff
@@ -39,14 +44,20 @@ require (
...
-	golang.org/x/crypto v0.0.0-20191002192127-34f69633bfdc
-	golang.org/x/net v0.0.0-20191002035440-2ec189313ef0
+	golang.org/x/crypto v0.0.0-20200622213623-75b288015ac9
+	golang.org/x/net v0.0.0-20200707034311-ab3426394381
 	golang.org/x/text v0.3.3 // indirect
-	golang.org/x/time v0.0.0-20180412165947-fbb02b2291d2
+	golang.org/x/time v0.0.0-20191024005414-555d28b269f0
+	golang.org/x/tools v0.0.0-20200806022845-90696ccdc692 // indirect
+	golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1 // indirect
+	google.golang.org/api v0.29.0 // indirect
```
[the remark this answers is not in the document]
Comment (ptabor) [reply go.mod:55]: (I'm sorry - I overlooked that my comments in review were 'pending' and I assumed they got submitted).

**Comment by lemonlinger** [comment]
Kindly ping~ How's the pr going? @skyao

**Comment by skyao** [comment]
@ptabor  shoud we continue for this issue of grpc version upgrade? I think we need to list what to change and also how.

**Comment by ptabor** [comment]
I'm not maintainer here, but my recommendation is following:

TLDR; Copy naming.Update class alone. Re-implement rest in terms of 'grpc.resolver' library.


1. The naming package: defines a format of key-value pairs that represent  a   (name ->  addr) binding. 
          The format is that under the key: `target+ "/"+addr`  we persist a json: 
            { Op: 'Add',  Addr: '192.168.0.15', Metadata: ''}
          it uses marshaling of the naming.Update format. 
    The API is public:  https://etcd.io/docs/v3.4.0/dev-guide/grpc_naming/     
        
2.  We need to assume some customers have etcd 3.4.0 instance and speak  to it using client libraries coming from etcd 3.5.0 and 3.4.0. There is no atomic switch of 'client libraries'. It might happen that some customer wrote (resolver.Update) using 3.4.0:

```
r.Update(context.TODO(), "my-service", naming.Update{Op: naming.Delete, Addr: "1.2.3.4"})
```

While they want to use etcd 3.5.0 client and use https://godoc.org/google.golang.org/grpc#WithResolvers to let grpc use etcd for naming resolution. 

3. Thus the format provided by: naming.Update{...} must not change, while the use-case of using of etcd as a naming resolution source should be preserved.  Of course the code that users [used to integrate the resolver](https://etcd.io/docs/v3.4.0/dev-guide/grpc_naming/#using-etcd-discovery-with-go-grpc)  from 3.5.0 will be different that the code that used to use with 3.4.0, but that's OK as @xiang90 pointed out.

> I think we can't split them: if we upgrade the grpc version and only fix one of them, then etcd can't compile.
I believe grpc 1.29.1 supports both: old and new interfaces. So I would assume we can do partial changes without bumping the grpc version, and bump when we are ready with all the changes.

**Comment by ptabor** [comment]
@skyao Do you plan to continue working on this issue ?

**Comment by skyao** [comment]
> @skyao Do you plan to continue working on this issue ?

I would like to continue this work, but now I'm comfused about how to continue...... Is there any conclusion?

**Comment by ptabor** [comment]
@skyao: I assumed that if we get the state from the recommendation, nobody would complain that its bad outcome. 
The main risk factor is me overlooking something, thus making this proposal infeasible. If you think, you can implement it, 
I would assume its the best verification. 

@xiang90 @gyuho  @jpbetz  Could you, please, review the recommendation from https://github.com/etcd-io/etcd/pull/12398#issuecomment-730456062 to unblock work on this issue ?

**Comment by skyao** [comment]
I would like to try my best to continue this topic of grpc version upgrade, but it seems the topic is too big to discuss and update code in ONE PR (and also this original PR is now have too many conflicting to current master branch). 

I will split this PR to some new and small PRs which will just focus in one specified issue in grpc version upgrade:

1. naming package is removed in grpc v1.3.0

     New PR is here: https://github.com/etcd-io/etcd/pull/12609

2. grpc loadbalancer api has some broken changes
3. we need to decide which versions should we upgrade to: including grpc / protobuf / gogo-protobuf ...

**Comment by ptabor** [comment]
Closing this one in favor of: https://github.com/etcd-io/etcd/issues/12652. 

If balancer alone change passes all the tests - we will need it.
